rcl_interfaces/msg/ParameterDescriptor.msg

```diff
@@ -1,8 +1,32 @@
 # This is the message to communicate a parameter's descriptor.
 
 string name
-uint8 type # Enum defined in ParameterType.msg
-# If 'true' then the value cannot change after it has been initialized
+
+# Enum values are defined in the `ParameterType.msg` message.
+uint8 type
+
+# Description of the parameter, visible from introspection tools.
+string description
+
+# Parameter constraints
+
+# Plain English description of additional constraints which cannot be expressed
+# with the available constraints, e.g. "only prime numbers".
+# By convention, this should only be used to clarify constraints which cannot
+# be completely expressed with the parameter constraints below.
+string additional_constraints
```

Comment:
I would argue the text can already be part of the description and doesn't have to be separate. E.g. a UI could show the description in a tooltip. I find it rather intuitive if it needs to concatenate two description to provide all information.

Reply:
The idea is that is only about constraints, and should be short. Where as the description would be longer and have information like what units it is in or how it used in the node.

I feel that it has value as a separate field. Consider that you're just using float range, you may still put in the text description that the parameter should be between 0.0 and 1.0. I don't see this as any different, you might want to speak about some of the constraints in the description, but repeat them more specifically or concisely in this field.

Also, I imagine that the GUI may not show the description all the time, perhaps only on hover, whereas this string would be shown always, maybe just under things like sliders or combo boxes as a footnote.

It would be nice to hear what others think about this, I did ask for feedback on it originally (https://github.com/ros2/rcl_interfaces/pull/73) and no one said anything.

Reply:
I think there is a semantic difference between the describing a parameter and describing the constraints on a parameter. So having two fields makes sense to me.

On another note, I could see this field being abused (for better or worse) to actually do constraint checking in user-defined callbacks. For example, [in the Navigation WG meeting today](https://discourse.ros.org/t/ros2-navigation-wg-thursday-3-00-pm-pacific-gmt-7-00/7586/16), there was a discussion about having a mechanism for constraining parameter updates in certain states of a lifecycle node. This string could be used to encode a valid state. I'm not sure if this is something we should be worried about though.

Reply:
I suppose they could abuse it that way, but they could also just put that into the description :x

That's only helpful, however, if the GUI understands it. So if you have cooperation with the GUI and code, why not just propose a new constraint for this message or just publish another message with your custom constraints...

---

W.r.t. the constraints being based on the lifecycle state, we should make it so that the ParameterDescriptor can be updated (currently you have to undeclare and re-declare a parameter to do that), so that they can simply update the constraints when the node itself changes states.

Reply:
ROS 1 has range limits as well as a description for dynamic parameters. That has worked well in the past without a second text field describing the constraint in text form. Without a strong reason I don't think we should add a second description just based on the thoughts about how it could useful.

First get a minimal version merged and implemented from APIs in C++ and Python, do constraint checking, and have an rqt plugin to interact with these parameters and then consider extending it if there is a strong need / use case to do so.

Reply:
I think the idea of how it should be used is sufficiently concrete to include it now, so I think we should keep it.

Reply:
To be clear here, I'm fine with removing it, if the consensus is to do that, but my personal opinion is that it should stay. Right now we've got a -1 from @dirk-thomas, +1 from me, and I'll call it a +0 from @jacobperron?

If anyone has a +/-1 opinion could comment here so we can have some more decisive conclusion would be nice.

Reply:
+1